core/control-plane/contract/src/main/java/org/eclipse/dataspaceconnector/contract/offer/ContractOfferServiceImpl.java

```diff
@@ -55,15 +57,21 @@ public ContractOfferServiceImpl(ParticipantAgentService agentService, ContractDe
     @NotNull
     public Stream<ContractOffer> queryContractOffers(ContractOfferQuery query, Range range) {
         var agent = agentService.createFor(query.getClaimToken());
-
-        return definitionService.definitionsFor(agent, range)
+        var offers = definitionService.definitionsFor(agent)
```

Comment:
I know that there aren't many ways to do this, but could this lead to long-blocked threads? 
I mean, for every page we have to fetch all the definitions - that could be not a big issue, also in the case if there are a lot of them, since they are really little.
But, on the last page request here the service will fetch all the definitions, query/evaluate all the policies, fetch all the assets, and then `skip` and `limit` are applied, as they are located after a `flatMap`, so the time needed to fetch the last page is actually the same time needed to fetch all the catalog in a single call.

e.g. (speculative) having 100.000 (pretty standard value) definitions means that there will be 200.000 queries to the policy store (access and contract) and 100.000 to the asset index. let's assume that the stores are really fast and the policy evaluation take no-time, if we count 1ms for every definition the single page request could take a time around 1 minute.

I don't remember if the issue with the "full catalog" fetch was with this operation of with serdes/transmission, but I think this implementation could lead poor performances and potentially to OOM exceptions.

Reply:
>... skip and limit are applied, as they are located after a flatMap
well, from a logical standpoint, that is the only place where `skip` and `limit` _can_ be applied, because we need the flat list of assets/offers first. 
Maybe we could collect first all the `Assets`, slice out the page and _then_ convert them into offers? That would still rely on the store not materializing the stream prematurely, but would save calls to the `PolicyDefinitionStore`.

Also, the way the `range` was applied before, was flat-out wrong, because it limited the number of `contractDefinitions`, so _some_ action needed to be taken.

The problem with the full list was that jetty ran into timeouts (and potentially body-size limits) when using anything other than in-mem.

> I mean, for every page we have to fetch all the definitions ...
We could add a cache for the definitions and policies (see #1173), if that needs to be optimized.

Reply:
The implementation was updated in commit https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/2018/commits/ea755913749a7e98e5c82891cd951fc46e8e7a25, as was the PR descriptions